### ISSUE TITLE
Performances decreased in IE between 0.7.x and 0.8.x

### ISSUE THREAD

**Issue by fnicollet** [mapbox/mapbox-gl-js#1336]
Hello,

I have noticed that the performances of mapbox-gl decreased between 0.7.x and 0.8.1. Very easy to reproduce, just any mapbox-gl map will do. When you drag the map around, you can feel a lag that wasn't there before (or not as much).
I am unsure when these issues started to happen as there has been many code changes between these two versions. Maybe because mapbox-gl now displays more labels (which is a good thing) ?
The performances also decreased in other browsers like Chrome but a little less noticeably

I am using Windows 8.1 x64 it that helps, tell me if you need more details

Fabien

**Comment by mourner** [comment]
My hunch would be to check commits 82ee08963822147cae86ddc6d17556143e7bfb71 and f5a34c61ef2e5423b77ca7ef5e4dc52a28ff7d39. If the first is fine but second is slow, then it's #1257.

**Comment by fnicollet** [comment]
Hello,

I think you have the first commit number wrong as https://github.com/mapbox/mapbox-gl-js/commit/82ee08963822147cae86ddc6d17556143e7bfb71 is about documentation. I tried reverting this one https://github.com/mapbox/mapbox-gl-js/commit/f5a34c61ef2e5423b77ca7ef5e4dc52a28ff7d39 but it doesn't seem to have any effect on performances. Not sure about #1257 as the performance are already inferior in the published 0.8.0 that I tried using
https://api.tiles.mapbox.com/mapbox-gl-js/v0.8.0/mapbox-gl.js

**Comment by mourner** [comment]
@fnicollet no, I mean the docs commit comes before a series of commits up to f5a34c6 which could be the cause. These come before 0.8.0.

**Comment by fnicollet** [comment]
Sorry, I will give it a try!

**Comment by fnicollet** [comment]
Just tried https://github.com/mapbox/mapbox-gl-js/commit/82ee08963822147cae86ddc6d17556143e7bfb71 and it is already slow. Will try digging further ...

**Comment by fnicollet** [comment]
Ok, nevermind, it looks like it comes from using leaflet-mapbox-gl https://github.com/mapbox/mapbox-gl-leaflet. Not sure why but it seems that you changed a few things related to navigation for 0.8. Is there something specific that could make using leaflet-mapbox-gl less "sync" that it was before?

Fabien

**Comment by mourner** [comment]
Then you'll have to do a `git bisect` to find the exact culprit.

**Comment by fnicollet** [comment]
Never tried it but it sound like an interesting thing to learn by looking at the docs, i'll give it a try!

**Comment by fnicollet** [comment]
Actually, my test was bad, this is not a leaflet-mapbox-gl problem. I tested in jsfiddle and the viewport was very small. So small that it wasn't lagging, probably because there was a lot less to render.
Just tried a map in "full screen" (not even in a crazy resolution) and it lags. 

I think it is a performance issue as the CPU goes up 50% usage when i just drag the map in already loaded areas. I would think of a labelling conflict issue but i'll have to "git bisect" I guess :)

Can some of you (windows machines anyone?) reproduce the same changes between 0.7 and 0.8?

Fabien

**Comment by mourner** [comment]
@fnicollet any updates?

**Comment by fnicollet** [comment]
Didn't have time yet to give "git bisect" a try, sorry, will do it asap

**Comment by mourner** [comment]
No rush, don't worry — I'm just checking :)

**Comment by fnicollet** [comment]
Before I try the git bisect operation, here are a few informations I got from the IE debugger "UI responsiveness". 

Here, I am trying with the simplest mapbox-gl map, no leaflet involved. I am using the outdoor style as this is the one which lags the most. My test is simple, I load the map and just pan around.

Basically, with 0.7.0, I can reach 10-15 FPS, with a 100% CPU usage. each animation frame take about 80ms:
![v0 7_ie](https://cloud.githubusercontent.com/assets/1240481/8556021/77e8d25e-24f4-11e5-8ae9-96575e5566ba.jpg)

When I use 0.8.1, I reach about 4 FPS. the CPU usage is lower (about 50%) but each animation frame take about 140ms:
![v0 8 1_ie](https://cloud.githubusercontent.com/assets/1240481/8556039/a0ced39e-24f4-11e5-9fff-a0aaa5f933aa.jpg)

Here is the profiler in 0.7:
![v0 7_profiler_ie](https://cloud.githubusercontent.com/assets/1240481/8556184/af8de48c-24f5-11e5-9ba0-e402d24309c2.jpg)

And in 0.8.1:
![v0 8 1_profiler_ie](https://cloud.githubusercontent.com/assets/1240481/8556188/b6b45e30-24f5-11e5-8bb6-e7091462c7c6.jpg)

In 0.8.1, the drawArrays and drawElements functions take about 3 times longer. It looks like this code is taking place outside of the Workers, so this would be what's eating the CPU I guess.

**Comment by fnicollet** [comment]
Just confirmed that the commit that made FPS drop is:
https://github.com/mapbox/mapbox-gl-js/commit/9289a2c069b68149899c1d987c716ad217e32cc0
It also produces weird rendering issues at that point, which have probably been fixed later:
![image](https://cloud.githubusercontent.com/assets/1240481/8558468/ae9feb88-2504-11e5-8aef-8ee022dd0be2.png)

**Comment by fnicollet** [comment]
I tried digging futher but the issues is pretty weird!

It looks like this line is what creates the glitches:
https://github.com/mapbox/mapbox-gl-js/commit/9289a2c069b68149899c1d987c716ad217e32cc0#diff-d19d663e570db606d5d3c4107a1dbed8R43

If I keep the util.inherit, even without adding the failIfMajorPerformanceCaveat property, I get the glitch:
```
var attributes = util.inherit(this._contextAttributes, {
        //failIfMajorPerformanceCaveat: failIfMajorPerformanceCaveat
    });
return this.canvas.getContext('experimental-webgl', attributes);
```
But if I pass the _contextAttributes directly, the rendering is all good:
```
    var attributes = util.inherit(this._contextAttributes, {
        //failIfMajorPerformanceCaveat: failIfMajorPerformanceCaveat
    });
	return this.canvas.getContext('experimental-webgl', this._contextAttributes);
```

This is like ... weird ...

**Comment by fnicollet** [comment]
The only difference I see is that attributes has a double-layer of _proto_:
![image](https://cloud.githubusercontent.com/assets/1240481/8559082/6a4d93a4-2509-11e5-87bc-4cc60d88b55b.png)

While _contextAttribute has only one _proto_:
![image](https://cloud.githubusercontent.com/assets/1240481/8559092/7f79bbae-2509-11e5-856f-a60fc6815793.png)

**Comment by mourner** [comment]
Wow, great investigation! Weird indeed,

**Comment by fnicollet** [comment]
It doesn't seem to make any difference if I apply the correction in 0.8 master though ... I am lost

**Comment by fnicollet** [comment]
Looks like this issue was fixed between 0.7.0 and 0.8.0 so i'll bisect further in time now

**Comment by fnicollet** [comment]
Well, it looks like you were right from the beginning @mourner ! 
Each of those commits: 
https://github.com/mapbox/mapbox-gl-js/pull/1257
Make the performances go worse and worse.

@lucaswoj did you notice some performances changes after this PR?

**Comment by lucaswoj** [comment]
@fnicollet We expected (and observed) a minor performance degradation from #1257 but nothing this bad. I will investigate the problem further this week. 

@mourner @fnicollet What browser, OS, and graphics card do you have?

**Comment by fnicollet** [comment]
@lucaswoj I run Windows 8.1 64 bits, with latest Chrome and latest IE 11. I have two graphic cards: HD Graphics 4000 + Nvidia GeForce GT740 M

How do you run the FPS benchmark that you did in 1257?

Thanks,

**Comment by lucaswoj** [comment]
@fnicollet to run the benchmark

 1. start the debugging server by running `npm start` in the repository's root directory
 2. navigate to `http://localhost:9966/bench/fps/?access_token=[YOUR MAPOBX ACCESS TOKEN]` in your browser
 3. keep the website in the foreground until the benchmarks finish (should only be two minutes or so)

cc @ansis @kkaefer this might be indicative of a fundamental problem with our approach to data driven styling

**Comment by lucaswoj** [comment]
Related to #1235 #1082

**Comment by lucaswoj** [comment]
I was able to replicate this problem on physical hardware. These results indicate this is regression likely affects all IE11 users and is not graphics card specific. 

### System Information

 - Windows 8.1 Pro
 - 8 GB RAM
 - Intel Core i7 processor @ 2.00 GHz
 - 64 bit
 - Intel HD Graphics 400 graphics card (1792 MB shared system memory)

### IE11
 
```
duration: 10000
13.3 fps /debug/mapbox-gl.js (eff02580a58ea4e55e41e93818e9c23b129a7126)
```

```
duration: 10000
14.5 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.8.1/mapbox-gl.js
14.2 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.8.0/mapbox-gl.js
45.8 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.7.0/mapbox-gl.js
```
 
### Chrome

```
duration: 10000
57.6 fps /debug/mapbox-gl.js (eff02580a58ea4e55e41e93818e9c23b129a7126)
```

```
duration: 10000
50.3 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.8.1/mapbox-gl.js
54.3 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.8.0/mapbox-gl.js
59.1 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.7.0/mapbox-gl.js
```

**Comment by lucaswoj** [comment]
So, if this is a result of #1257 (which I think is likely) then we don't have many options for improving performance right now. 

We could *maybe* eliminate a few redundant `gl.disableVertexAttribArray` calls but my hunch is that won't help much. 

The only way I could imagine moving away from `gl.disableVertexAttribArray` and `gl.vertexAttrib*` calls is to implement dynamic shader code generation (which is vaguely on the table in #1318 #1356)

**Comment by ansis** [comment]
Is [`bindAttribLocation`](https://github.com/mapbox/mapbox-gl-js/blob/eff02580a58ea4e55e41e93818e9c23b129a7126/js/render/gl_util.js#L53) working properly? maybe it's still disabling attribute 0 which has negative performance impact.

I think it's worth trying to eliminate unnecessary attrib array enables/disables (https://github.com/mapbox/mapbox-gl-js/commit/1c34f178776a0dc1c78a2e613c4a8f0ae6896124) by keeping track of attribute state in js. It would be a bit weird if this fixed it, but I wouldn't be *that* surprised.

and triple check that 82ee08963822147cae86ddc6d17556143e7bfb71 is still fast

**Comment by mourner** [comment]
I'm also really curious to see the performance of Project Spartan. Given that IE11 already has a lot of performance quirks (e.g. no Transferable Objects support, leading to stuttering on tile loads), it may as well happen that we'll have no choice but to give up fast IE11 support and just wait for the next version.

**Comment by ansis** [comment]
> no Transferable Objects support

I haven't tested, but supposedly IE11 supports one transferable object: https://github.com/mapbox/mapbox-gl-js/issues/95#issuecomment-51842674

**Comment by mourner** [comment]
Supposedly, but I still haven't seen any practical evidence for that :)

**Comment by fnicollet** [comment]
thanks everyone for looking into this :) and thanks @lucaswoj for the instructions. Here are my results in IE:
before commits (https://github.com/mapbox/mapbox-gl-js/commit/82ee08963822147cae86ddc6d17556143e7bfb71)
30.6 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.7.0/mapbox-gl.js
27.5 fps /debug/mapbox-gl.js
31.2 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.7.0/mapbox-gl.js
27.8 fps /debug/mapbox-gl.js

after commits (https://github.com/mapbox/mapbox-gl-js/commit/f5a34c61ef2e5423b77ca7ef5e4dc52a28ff7d39)
29 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.7.0/mapbox-gl.js
10.4 fps /debug/mapbox-gl.js
31.2 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.7.0/mapbox-gl.js
10.3 fps /debug/mapbox-gl.js

**Comment by lucaswoj** [comment]
@fnicollet Good news! We removed all the offending code from the v8 release. Hopefully we can figure out how to make data-driven styling fast in all browsers in time for v9!

**Comment by fnicollet** [comment]
@lucaswoj Cool, that sounds good :). I can't see the revert commits in master, did it not have time to be merged before 0.9?

**Comment by lucaswoj** [comment]
@fnicollet Unfortunately not in time for `0.9`. Sitting in a PR right now https://github.com/mapbox/mapbox-gl-js/pull/1418
